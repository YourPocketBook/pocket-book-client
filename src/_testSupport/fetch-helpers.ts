import fetchMock, {
  MockOptionsMethodDelete,
  MockOptionsMethodGet,
  MockOptionsMethodPost,
  MockOptionsMethodPut,
  MockResponse,
  MockResponseFunction
} from "fetch-mock";

afterEach(fetchMock.reset);

export function authenticatedGet(
  url: string,
  token: string,
  response: MockResponse | MockResponseFunction,
  options?: MockOptionsMethodGet
) {
  fetchMock.get(
    (u, h) =>
      u === url &&
      (h.headers as { Authorization: string }).Authorization ===
        `Bearer ${token}`,
    response,
    options
  );
}

export function authenticatedPut(
  url: string,
  token: string,
  body: {},
  response: MockResponse | MockResponseFunction,
  options?: MockOptionsMethodPut
) {
  fetchMock.put(
    (u, h) =>
      u === url &&
      (h.headers as { Authorization: string }).Authorization ===
        `Bearer ${token}` &&
      h.body === JSON.stringify(body),
    response,
    options
  );
}

export function authenticatedPost(
  url: string,
  token: string,
  body: {},
  response: MockResponse | MockResponseFunction,
  options?: MockOptionsMethodPost
) {
  fetchMock.post(
    (u, h) =>
      u === url &&
      (h.headers as { Authorization: string }).Authorization ===
        `Bearer ${token}` &&
      h.body === JSON.stringify(body),
    response,
    options
  );
}

export function authenticatedDelete(
  url: string,
  token: string,
  response: MockResponse | MockResponseFunction,
  options?: MockOptionsMethodDelete
) {
  fetchMock.delete(
    (u, h) =>
      u === url &&
      (h.headers as { Authorization: string }).Authorization ===
        `Bearer ${token}`,
    response,
    options
  );
}
