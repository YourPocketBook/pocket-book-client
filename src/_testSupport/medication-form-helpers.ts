import { fireEvent } from "@testing-library/react";

export const testDefaultState = {
  adviceIfDeclined: "Test Advice 1",
  adviceIfTaken: "Test Advice 2",
  dose: "Test Dose",
  exclusionCriteria: "Test Exclusion",
  form: "Test Form",
  inclusionCriteria: "Test Inclusion",
  indications: "Test Indications",
  name: "Test Name",
  policyDate: "2019-12-10",
  route: "Test Route",
  sideEffects: "Test Side Effects"
};

export function setupState(getByLabelText: (text: string) => HTMLElement) {
  fireEvent.change(getByLabelText("Name"), {
    target: { value: testDefaultState.name }
  });
  fireEvent.change(getByLabelText("Policy Date"), {
    target: { value: testDefaultState.policyDate }
  });
  fireEvent.change(getByLabelText("Indications"), {
    target: { value: testDefaultState.indications }
  });
  fireEvent.change(getByLabelText("Inclusion Criteria"), {
    target: { value: testDefaultState.inclusionCriteria }
  });
  fireEvent.change(getByLabelText("Exclusion Criteria"), {
    target: { value: testDefaultState.exclusionCriteria }
  });
  fireEvent.change(getByLabelText("Name, Forms and Strengths of Medicine"), {
    target: { value: testDefaultState.form }
  });
  fireEvent.change(getByLabelText("Route or Method"), {
    target: { value: testDefaultState.route }
  });
  fireEvent.change(getByLabelText("Dosage"), {
    target: { value: testDefaultState.dose }
  });
  fireEvent.change(getByLabelText("Side Effects"), {
    target: { value: testDefaultState.sideEffects }
  });
  fireEvent.change(getByLabelText("Advice if Taken"), {
    target: { value: testDefaultState.adviceIfTaken }
  });
  fireEvent.change(getByLabelText("Advice if Declined"), {
    target: { value: testDefaultState.adviceIfDeclined }
  });
}

export function checkForClass(
  getByLabelText: (text: string) => HTMLElement,
  cls: "is-valid" | "is-invalid"
) {
  expect(getByLabelText("Name")).toHaveClass(cls);
  expect(getByLabelText("Policy Date")).toHaveClass(cls);
  expect(getByLabelText("Indications")).toHaveClass(cls);
  expect(getByLabelText("Inclusion Criteria")).toHaveClass(cls);
  expect(getByLabelText("Exclusion Criteria")).toHaveClass(cls);
  expect(getByLabelText("Name, Forms and Strengths of Medicine")).toHaveClass(
    cls
  );
  expect(getByLabelText("Route or Method")).toHaveClass(cls);
  expect(getByLabelText("Dosage")).toHaveClass(cls);
  expect(getByLabelText("Side Effects")).toHaveClass(cls);
  expect(getByLabelText("Advice if Taken")).toHaveClass(cls);
  expect(getByLabelText("Advice if Declined")).toHaveClass(cls);
}

export function checkState(
  getByLabelText: (text: string) => HTMLElement,
  hasValue: boolean
) {
  expect(getByLabelText("Name")).toHaveValue(
    hasValue ? testDefaultState.name : ""
  );
  expect(getByLabelText("Policy Date")).toHaveValue(
    hasValue ? testDefaultState.policyDate : ""
  );
  expect(getByLabelText("Indications")).toHaveTextContent(
    hasValue ? testDefaultState.indications : ""
  );
  expect(getByLabelText("Inclusion Criteria")).toHaveTextContent(
    hasValue ? testDefaultState.inclusionCriteria : ""
  );
  expect(getByLabelText("Exclusion Criteria")).toHaveTextContent(
    hasValue ? testDefaultState.exclusionCriteria : ""
  );
  expect(
    getByLabelText("Name, Forms and Strengths of Medicine")
  ).toHaveTextContent(hasValue ? testDefaultState.form : "");
  expect(getByLabelText("Route or Method")).toHaveTextContent(
    hasValue ? testDefaultState.route : ""
  );
  expect(getByLabelText("Dosage")).toHaveTextContent(
    hasValue ? testDefaultState.dose : ""
  );
  expect(getByLabelText("Side Effects")).toHaveTextContent(
    hasValue ? testDefaultState.sideEffects : ""
  );
  expect(getByLabelText("Advice if Taken")).toHaveTextContent(
    hasValue ? testDefaultState.adviceIfTaken : ""
  );
  expect(getByLabelText("Advice if Declined")).toHaveTextContent(
    hasValue ? testDefaultState.adviceIfDeclined : ""
  );
}
