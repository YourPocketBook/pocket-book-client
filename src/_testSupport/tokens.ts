export const inDateNonAdminToken =
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwiaWF0IjoxNDE2MjM5MDIyLCJleHAiOjE3MTYyNDAwMjIsImF1ZCI6ImxvY2FsaG9zdCIsImlzcyI6ImxvY2FsaG9zdCIsImdpdmVuX25hbWUiOiJUZXN0IFVzZXIiLCJuYW1laWQiOiJmZjFkMDhlNy00NjU2LTRjMjgtOTc2Zi0xYzI3ZjE5MTQ3N2YiLCJlbWFpbCI6InRlc3RAc2phLm9yZy51ayJ9.VmT5YkTG0sUd5sK7skfMdi9kc5XG7Yhx6sqhqMSxhVY";
export const inDateAdminToken =
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwiaWF0IjoxNDE2MjM5MDIyLCJleHAiOjE3MTYyNDAwMjIsImF1ZCI6ImxvY2FsaG9zdCIsImlzcyI6ImxvY2FsaG9zdCIsInJvbGUiOiJBZG1pbiIsImdpdmVuX25hbWUiOiJUZXN0IFVzZXIiLCJuYW1laWQiOiJmZjFkMDhlNy00NjU2LTRjMjgtOTc2Zi0xYzI3ZjE5MTQ3N2YiLCJlbWFpbCI6InRlc3RAc2phLm9yZy51ayJ9.rSQb2Hv6RfxaXSfO6s3QjLjnYKeqVgps3rPenquh3Nk";
export const expiredToken =
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwiaWF0IjoxNDE2MjM5MDIyLCJleHAiOjE0MTYyNDAwMjIsImF1ZCI6ImxvY2FsaG9zdCIsImlzcyI6ImxvY2FsaG9zdCIsInVuaXF1ZV9uYW1lIjoiVGVzdCBVc2VyIiwibmFtZWlkIjoiZmYxZDA4ZTctNDY1Ni00YzI4LTk3NmYtMWMyN2YxOTE0NzdmIiwiZW1haWwiOiJ0ZXN0QHNqYS5vcmcudWsifQ.7bDfHDk9sWXL8boFeZWIdo_GB-4Qj8PvA4HDC49EzNc";
