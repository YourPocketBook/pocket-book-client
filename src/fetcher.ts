import isFuture from "date-fns/isFuture";
import jwtDecode from "jwt-decode";

class Fetcher {
  private realName: string = "";
  private nameChanged?: (name: string) => void;
  private token: string | null;
  private parsedToken:
    | { exp: number; given_name: string; role: string }
    | undefined;

  constructor() {
    this.token = localStorage.getItem("token");
    this.parseToken();
  }

  public setNameChangeHandler(func?: (name: string) => void) {
    this.nameChanged = func;
  }

  public authFetch(input: RequestInfo, init?: RequestInit): Promise<Response> {
    if (!this.token) {
      throw new Error("No token");
    }

    init = {
      ...init,
      headers: {
        ...(init ? init.headers : {}),
        Authorization: `Bearer ${this.token}`
      }
    };

    return fetch(input, init);
  }

  public saveToken(token: string) {
    localStorage.setItem("token", token);
    this.token = token;
    this.parseToken();
  }

  public checkToken(token: string): boolean {
    return this.token === token;
  }

  public clearToken() {
    localStorage.removeItem("token");
    this.token = null;
    this.parsedToken = undefined;
  }

  public getName() {
    return this.realName;
  }

  public updateName(name: string) {
    this.realName = name;
    if (this.nameChanged) {
      this.nameChanged(name);
    }
  }

  public isAdmin() {
    if (this.parsedToken) {
      return this.parsedToken.role === "Admin";
    }
    return false;
  }

  public hasToken(): boolean {
    return !!this.token;
  }

  public isInDate() {
    if (this.parsedToken) {
      return isFuture(this.parsedToken.exp * 1000);
    }

    return false;
  }

  private parseToken() {
    if (this.token) {
      try {
        this.parsedToken = jwtDecode(this.token);
      } catch {
        this.token = null;
        this.updateName("");
        localStorage.removeItem("token");
        this.parsedToken = undefined;
      }

      if (this.parsedToken) {
        this.updateName(this.parsedToken.given_name);
      }
    } else {
      this.parsedToken = undefined;
      this.updateName("");
    }
  }
}

const fetcher = new Fetcher();

export { fetcher, Fetcher };
