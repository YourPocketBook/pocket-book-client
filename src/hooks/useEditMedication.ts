import isValid from "date-fns/isValid";
import parseISO from "date-fns/parseISO";
import { useReducer } from "react";
import {
  IEditMedication,
  IEditMedicationValid
} from "../medications/medication-form";

export function useEditMedication() {
  const [state, dispatch] = useReducer(
    (
      s: IEditMedication & { valid: IEditMedicationValid },
      a: { prop: keyof IEditMedication } & (
        | { op: "update"; value: string | undefined | null }
        | {
            op: "invalidate";
          }
        | { op: "setValid"; value: boolean }
      )
    ) => {
      if (s.hasOwnProperty(a.prop)) {
        const newData = { ...s };

        switch (a.op) {
          case "update":
            newData[a.prop] = a.value || "";
            if (a.prop === "policyDate") {
              newData.valid[a.prop] = !!a.value && isValid(parseISO(a.value));
            } else {
              newData.valid[a.prop] = !!a.value && a.value.length > 0;
            }
            break;
          case "invalidate":
            newData.valid[a.prop] = false;
            break;
          case "setValid":
            newData.valid[a.prop] = a.value;
            break;
        }
        return newData;
      }
      return s;
    },
    {
      adviceIfDeclined: "",
      adviceIfTaken: "",
      dose: "",
      exclusionCriteria: "",
      form: "",
      inclusionCriteria: "",
      indications: "",
      name: "",
      policyDate: "",
      route: "",
      sideEffects: "",
      valid: {
        adviceIfDeclined: false,
        adviceIfTaken: false,
        dose: false,
        exclusionCriteria: false,
        form: false,
        inclusionCriteria: false,
        indications: false,
        name: false,
        policyDate: false,
        route: false,
        sideEffects: false
      }
    }
  );

  return { state, dispatch };
}
