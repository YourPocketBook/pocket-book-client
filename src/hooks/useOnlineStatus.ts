// Included from https://github.com/21kb/react-hooks/blob/master/packages/react-online-status-hook
//
// Copyright Â© 2018 Kunall Banerjee
// MIT License

import { useEffect, useState } from "react";

export function useOnlineStatus() {
  const [state, setState] = useState(navigator.onLine);

  const onOnlineEvent = () => {
    setState(navigator.onLine);
  };

  const onOfflineEvent = () => {
    setState(navigator.onLine);
  };

  useEffect(() => {
    window.addEventListener("online", onOnlineEvent);
    window.addEventListener("offline", onOfflineEvent);

    return () => {
      window.removeEventListener("online", onOnlineEvent);
      window.removeEventListener("offline", onOfflineEvent);
    };
  });

  return state;
}
