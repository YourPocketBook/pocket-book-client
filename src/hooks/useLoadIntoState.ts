import { Dispatch, useEffect, useState } from "react";
import { fetcher } from "../fetcher";

/**
 * Fetches JSON from the specified uri and loads each property into individually editable state variables
 *
 * @param uri The URI to load the data from
 * @param setState An object containing setState functions for each property
 * @param [authenticate=false] If set to true, the current token will be passed with the call
 * @returns True if the request is currently running, otherwise false.
 */
export function useLoadIntoEditableState<
  TData extends { [key: string]: string | undefined | null } & object
>(
  uri: string,
  dispatch: Dispatch<{
    op: "update";
    prop: keyof TData;
    value: string | undefined | null;
  }>,
  authenticate = false
): boolean {
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    const fetchAbort = new AbortController();

    const options: RequestInit = {
      cache: "reload",
      method: "GET",
      signal: fetchAbort.signal
    };

    async function func() {
      try {
        const res = authenticate
          ? await fetcher.authFetch(uri, options)
          : await fetch(uri, options);

        if (res.ok) {
          const data = (await res.json()) as TData;
          for (const prop in data) {
            if (data.hasOwnProperty(prop)) {
              dispatch({ op: "update", prop, value: data[prop] });
            }
          }
        }
      } catch (ex) {
        if (!fetchAbort.signal.aborted) {
          // tslint:disable-next-line:no-console
          console.log(`Could not get from ${uri}.`);
        }
      } finally {
        if (!fetchAbort.signal.aborted) {
          setLoading(false);
        }
      }
    }

    func();
    return () => fetchAbort.abort();
  }, [uri, authenticate, dispatch]);

  return loading;
}
