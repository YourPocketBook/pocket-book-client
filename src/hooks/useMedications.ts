import { useEffect, useState } from "react";
import { fetcher } from "../fetcher";
import { IMedication } from "../model/medication";
import { useOnlineStatus } from "./useOnlineStatus";

const medicationsKey = "medications";
const listEndpoint = "/api/get-medications";
const detailEndpoint = "/api/get-medication";

export function useMedicationList() {
  const [isLoading, setIsLoading] = useState(false);
  const [serverError, setServerError] = useState(false);
  const [updateAvailable, setUpdateAvailable] = useState(false);
  const isOnline = useOnlineStatus();

  const localStoreString = localStorage.getItem(medicationsKey);
  const localData = localStoreString
    ? (JSON.parse(localStoreString) as Array<Partial<IMedication>>)
    : [];
  const [medicationList, setMedicationList] = useState(localData);

  useEffect(() => {
    if (!isOnline) {
      return;
    }

    const fetchAbort = new AbortController();

    const options = {
      method: "GET",
      signal: fetchAbort.signal
    };

    setIsLoading(true);

    async function func() {
      try {
        const res = await fetcher.authFetch(listEndpoint, options);

        if (fetchAbort.signal.aborted) {
          return;
        }

        if (res.ok) {
          const data = (await res.json()) as Array<Partial<IMedication>>;

          if (fetchAbort.signal.aborted) {
            return;
          }

          const cacheString = localStorage.getItem(medicationsKey);
          const oldData = cacheString
            ? (JSON.parse(cacheString) as Array<Partial<IMedication>>)
            : [];

          const currentIds = data
            .filter(m => m && m.id)
            .map(m => m.id as number);
          const previousIds = oldData
            .filter(m => m && m.id)
            .map(m => m.id as number);

          const updatedItems = oldData
            .filter(m => currentIds.includes(m.id as number))
            .map(m => {
              const update = data.find(n => n.id === m.id);
              if (update && update.name) {
                m.name = update.name;
              }
              return m;
            })
            .concat(data.filter(m => !previousIds.includes(m.id as number)));

          setMedicationList(updatedItems);
          localStorage.setItem(medicationsKey, JSON.stringify(updatedItems));
          setUpdateAvailable(true);
          setServerError(false);
        } else {
          setServerError(true);
        }
      } catch (ex) {
        if (!fetchAbort.signal.aborted) {
          setServerError(true);
        }
      } finally {
        if (!fetchAbort.signal.aborted) {
          setIsLoading(false);
        }
      }
    }

    func();

    return () => fetchAbort.abort();
  }, [isOnline]);

  return {
    isLoading,
    medications: medicationList
      .filter(m => m && m.id)
      .map(m => ({
        availableOffline: !!m.inclusionCriteria,
        id: m.id as number,
        name: m.name as string
      })),
    serverError,
    updateAvailable
  };
}

export function useMedication(id: string, refresh: boolean) {
  const [isLoading, setIsLoading] = useState(false);
  const [updateAvailable, setUpdateAvailable] = useState(false);
  const [loadFailed, setLoadFailed] = useState(false);
  const isOnline = useOnlineStatus();

  const localStoreString = localStorage.getItem(medicationsKey);
  const data = localStoreString
    ? (JSON.parse(localStoreString) as Array<Partial<IMedication>>)
    : undefined;
  const med = data ? data.find(m => m.id === parseInt(id, 10)) : undefined;
  const [medication, setMedication] = useState(med);

  useEffect(() => {
    if (!isOnline) {
      return;
    }

    const fetchAbort = new AbortController();

    const options = {
      cache: refresh ? ("no-cache" as "no-cache") : undefined,
      method: "GET",
      signal: fetchAbort.signal
    };

    setIsLoading(true);
    setLoadFailed(false);

    async function func() {
      try {
        const res = await fetcher.authFetch(`${detailEndpoint}/${id}`, options);

        if (fetchAbort.signal.aborted) {
          return;
        }

        if (res.ok) {
          const newData = (await res.json()) as IMedication;

          if (fetchAbort.signal.aborted) {
            return;
          }

          setMedication(newData);
          cacheMedication(newData);
          setUpdateAvailable(true);
        } else {
          setLoadFailed(true);
        }
      } catch (ex) {
        if (!fetchAbort.signal.aborted) {
          // tslint:disable-next-line:no-console
          console.log(`Could not get from ${detailEndpoint}/${id}`);
          setLoadFailed(true);
        }
      } finally {
        if (!fetchAbort.signal.aborted) {
          setIsLoading(false);
        }
      }
    }

    func();

    return () => fetchAbort.abort();
  }, [isOnline, id, refresh]);

  return {
    isLoading,
    loadFailed,
    medication,
    updateAvailable
  };
}

export function cacheMedication(medication: IMedication) {
  const localStoreString = localStorage.getItem(medicationsKey);

  const data = localStoreString
    ? (JSON.parse(localStoreString) as Array<Partial<IMedication>>)
    : undefined;

  let newList: Array<Partial<IMedication>>;

  if (data) {
    newList = data.map(m => (m.id === medication.id ? medication : m));
  } else {
    newList = [medication];
  }

  localStorage.setItem(medicationsKey, JSON.stringify(newList));
}
