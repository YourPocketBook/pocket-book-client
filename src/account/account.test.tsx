import "@testing-library/jest-dom/extend-expect";
import { act, fireEvent, render, wait } from "@testing-library/react";
import fetchMock from "fetch-mock";
import React from "react";
import ReactDOM from "react-dom";
import { Account } from ".";
import {
  authenticatedGet,
  authenticatedPut
} from "../_testSupport/fetch-helpers";
import { inDateNonAdminToken } from "../_testSupport/tokens";
import { fetcher } from "../fetcher";
import { history } from "../history";
import { useOnlineStatus } from "../hooks/useOnlineStatus";

jest.mock("../hooks/useOnlineStatus");

const getEmailSettingsEndPoint = "/api/get-email-settings";
const setEmailSettingsEndPoint = "/api/update-email-settings";

describe("Account Page", () => {
  beforeAll(() => {
    fetcher.saveToken(inDateNonAdminToken);
  });

  beforeEach(() => {
    history.push = jest.fn();
    (useOnlineStatus as jest.Mock).mockReturnValue(true);
  });

  it("renders without crashing", () => {
    const div = document.createElement("div");

    ReactDOM.render(<Account />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it("renders correctly", async () => {
    authenticatedGet(getEmailSettingsEndPoint, inDateNonAdminToken, {
      updateEmailConsentGiven: true
    });

    const { container } = render(<Account />);

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(container).toMatchSnapshot();
  });

  it("displays correct email consent given correctly", async () => {
    authenticatedGet(getEmailSettingsEndPoint, inDateNonAdminToken, {
      updateEmailConsentGiven: true
    });

    const { getByText } = render(<Account />);

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(getByText("Disable Update Emails")).toBeInTheDocument();
  });

  it("displays correct email consent not given correctly", async () => {
    authenticatedGet(getEmailSettingsEndPoint, inDateNonAdminToken, {
      updateEmailConsentGiven: false
    });

    const { getByText } = render(<Account />);

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(getByText("Enable Update Emails")).toBeInTheDocument();
  });

  it("displays correct email consent given correctly when offline", async () => {
    authenticatedGet(getEmailSettingsEndPoint, inDateNonAdminToken, {
      updateEmailConsentGiven: true
    });

    const { getByText, rerender } = render(<Account />);

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    act(() => {
      (useOnlineStatus as jest.Mock).mockReturnValue(false);
    });

    rerender(<Account />);

    expect(getByText("Disable Update Emails")).toBeDisabled();
  });

  it("displays correct email consent not given correctly when offline", async () => {
    authenticatedGet(getEmailSettingsEndPoint, inDateNonAdminToken, {
      updateEmailConsentGiven: false
    });

    const { getByText, rerender } = render(<Account />);

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    act(() => {
      (useOnlineStatus as jest.Mock).mockReturnValue(false);
    });

    rerender(<Account />);

    expect(getByText("Enable Update Emails")).toBeDisabled();
  });

  it("hides email consent button when network fails", async () => {
    authenticatedGet(getEmailSettingsEndPoint, inDateNonAdminToken, {
      throws: new Error()
    });

    const { queryByText } = render(<Account />);

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(queryByText("Enable Update Emails")).toBeNull();
    expect(queryByText("Disable Update Emails")).toBeNull();
  });

  it("hides email consent button when network returns failure", async () => {
    authenticatedGet(getEmailSettingsEndPoint, inDateNonAdminToken, {
      status: 401
    });

    const { queryByText } = render(<Account />);

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(queryByText("Enable Update Emails")).toBeNull();
    expect(queryByText("Disable Update Emails")).toBeNull();
  });

  it("removes email consent when clicked", async () => {
    authenticatedGet(getEmailSettingsEndPoint, inDateNonAdminToken, {
      updateEmailConsentGiven: true
    });
    authenticatedPut(
      setEmailSettingsEndPoint,
      inDateNonAdminToken,
      { updateEmailConsentGiven: false },
      { status: 204 }
    );

    const { getByText } = render(<Account />);

    await wait(() =>
      expect(fetchMock.done(getEmailSettingsEndPoint)).toBeTruthy()
    );

    fireEvent.click(getByText("Disable Update Emails"));

    await wait(() => expect(fetchMock.done()).toBeTruthy());
  });

  it("grants email consent when clicked", async () => {
    authenticatedGet(getEmailSettingsEndPoint, inDateNonAdminToken, {
      body: { updateEmailConsentGiven: false }
    });
    authenticatedPut(
      setEmailSettingsEndPoint,
      inDateNonAdminToken,
      { updateEmailConsentGiven: true },
      { status: 204 }
    );

    const { getByText } = render(<Account />);

    await wait(() =>
      expect(fetchMock.done(getEmailSettingsEndPoint)).toBeTruthy()
    );

    fireEvent.click(getByText("Enable Update Emails"));

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(getByText("Disable Update Emails")).toBeInTheDocument();
  });

  it("handles email consent throwing properly", async () => {
    authenticatedGet(getEmailSettingsEndPoint, inDateNonAdminToken, {
      updateEmailConsentGiven: false
    });
    authenticatedPut(
      setEmailSettingsEndPoint,
      inDateNonAdminToken,
      { updateEmailConsentGiven: true },
      { throws: new Error() }
    );

    const { getByText } = render(<Account />);

    await wait(() =>
      expect(fetchMock.done(getEmailSettingsEndPoint)).toBeTruthy()
    );

    fireEvent.click(getByText("Enable Update Emails"));

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(getByText("Enable Update Emails")).toBeInTheDocument();
  });

  it("handles email consent failing properly", async () => {
    authenticatedGet(getEmailSettingsEndPoint, inDateNonAdminToken, {
      updateEmailConsentGiven: false
    });
    authenticatedPut(
      setEmailSettingsEndPoint,
      inDateNonAdminToken,
      { updateEmailConsentGiven: true },
      { status: 401 }
    );

    const { getByText } = render(<Account />);

    await wait(() =>
      expect(fetchMock.done(getEmailSettingsEndPoint)).toBeTruthy()
    );

    fireEvent.click(getByText("Enable Update Emails"));

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(getByText("Enable Update Emails")).toBeInTheDocument();
  });
});
