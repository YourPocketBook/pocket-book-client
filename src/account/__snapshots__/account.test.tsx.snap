// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account Page renders correctly 1`] = `
<div>
  <div
    class="background"
  >
    <nav
      class="navbar navbar"
    >
      <a
        class="navbar-brand"
      >
        PocketBook
      </a>
      <span
        class="navbar-text"
      >
        Hi, 
        Test User
      </span>
    </nav>
    <div
      class="centreContainer"
    >
      <div
        class="container"
      >
        <div
          class="row"
        >
          <div
            class="col-lg-6 offset-lg-3"
          >
            <div
              class="dashboardBackground"
            >
              <div
                class="container"
              >
                <div
                  class="row"
                >
                  <div
                    class="col"
                  >
                    <h1
                      class="text-center"
                    >
                      Account
                    </h1>
                  </div>
                </div>
                <div
                  class="dashboardButton row"
                >
                  <div
                    class="col"
                  >
                    <button
                      class="btn btn-dark-green btn-lg btn-block"
                      type="button"
                    >
                      Change Name
                    </button>
                  </div>
                </div>
                <div
                  class="dashboardButton row"
                >
                  <div
                    class="col"
                  >
                    <button
                      class="btn btn-dark-green btn-lg btn-block"
                      type="button"
                    >
                      Change Password
                    </button>
                  </div>
                </div>
                <div
                  class="dashboardButton row"
                >
                  <div
                    class="col"
                  >
                    <button
                      class="btn btn-dark-green btn-lg btn-block"
                      type="button"
                    >
                      Disable Update Emails
                    </button>
                  </div>
                </div>
                <div
                  class="dashboardButton row"
                >
                  <div
                    class="col"
                  >
                    <button
                      class="btn btn-danger btn-lg btn-block"
                      type="button"
                    >
                      Delete Account
                    </button>
                  </div>
                </div>
                <div
                  class="dashboardButton row"
                >
                  <div
                    class="col"
                  >
                    <a
                      class="btn btn-dark-green btn-block btn-lg"
                      href="/home"
                    >
                      Home
                    </a>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="col"
          />
          <div
            class="col"
          />
          <div
            class="col"
          />
        </div>
      </div>
    </div>
    <nav
      class="fixed-bottom footer navbar"
    >
      <span
        class="navbar-text"
      >
        Design Copyright Â© 
        2020
         Tony Richards.
      </span>
      <ul
        class="nav"
      >
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/privacy"
          >
            Privacy Policy
          </a>
        </li>
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/tsandcs"
          >
            Terms of Use
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;
